src/java.net.http/share/classes/jdk/internal/net/http/quic/QuicEndpoint.java

```diff
@@ -0,0 +1,2040 @@
+                if (debug.on()) debug.log("endpoint resumed reading in same virtual thread");
+            } // if readingStalled was false and readingDone was false there is nothing to do.
+        }
+
+        @Override
+        void pauseReading() {
+            boolean paused = false;
+            synchronized (this) {
+                if (bufferTooBig()) paused = readingStalled = true;
+            }
+            if (paused) {
+                // System.out.println(this + " endpoint paused reading");
+                if (debug.on()) debug.log("endpoint paused reading");
+            }
+        }
+
+        @Override
+        public ChannelType channelType() {
+            return BLOCKING_WITH_VIRTUAL_THREADS;
+        }
+
+        void attach(QuicVirtualThreadPoller poller) {
+            this.poller = poller;
+            var future = poller.startReading(this);
+            synchronized (this) {
+                this.key = future;
+            }
+        }
+
+        Executor writeLoopExecutor() {
+            QuicVirtualThreadPoller poller = this.poller;
+            if (poller == null) return executor;
+            return poller.readLoopExecutor();
+        }
+
+        private final SequentialScheduler channelScheduler = SequentialScheduler.lockingScheduler(this::channelReadLoop0);
+
+        @Override
+        void channelReadLoop() {
+            channelScheduler.runOrSchedule();
+        }
+
+        private void channelReadLoop0() {
+            super.channelReadLoop();
+        }
+
+        @Override
+        public void detach() {
+            var key = this.key;
+            try {
+                if (key != null) {
+                    // do not interrupt the reading task if running:
+                    // closing the channel later on will ensure that the
+                    // task eventually terminates.
+                    key.cancel(false);
+                }
+            } catch (Throwable e) {
+                if (debug.on()) {
+                    debug.log("Failed to cancel future: " + e);
+                }
+            }
+        }
+    }
+
+    private ByteBuffer copyOnHeap(ByteBuffer buffer) {
+        ByteBuffer onHeap = ByteBuffer.allocate(buffer.remaining());
+        return onHeap.put(buffer).flip();
+    }
+
+    void channelReadLoop() {
+        // we can't prevent incoming datagram from being received
+        // at this level of the stack. If there is a datagram available,
+        // we must read it immediately and put it in the read queue.
+        // Flow control, which is implemented at the higher level,
+        // should ensure that the queue does not overflow.
+        //
+        // At the moment we have a single endpoint, a single
+        // channel, and a single selector thread, so we can do
+        // the reading directly in the selector thread and offload
+        // the parsing (the readLoop) to the executor.
```

Comment:
This comment is outdated. We actually stop reading from the channel when the amount of data buffered exceeds a high watermark threshold.

Reply:
Updated in latest commits